core/api_handler.py

```diff
@@ -175,6 +167,107 @@ def upload_sequencing_run(sequencing_run):
     return run_id
 
 
+def _run_upload_handler(api_instance, sequencing_run, run_id):
+    """
+    Gets the multithreading config option and runs either the standard upload, or the multithreaded upload
+    :param api_instance: api instance should have already been initialized
+    :param sequencing_run: sequencing run to upload
+    :param run_id: run id to use for upload
+    :return:
+    """
+    multithreading_config = config.read_config_option('multithreading', bool, False)
+    threads_config = config.read_config_option('threads')
+    if multithreading_config is not None and multithreading_config is True and \
+            threads_config is not None and int(threads_config) > 1:
+        return _run_upload_threadpool(api_instance, sequencing_run, run_id, int(threads_config))
+    else:
+        return _run_upload_basic(api_instance, sequencing_run, run_id)
+
+
+def _run_upload_basic(api_instance, sequencing_run, run_id):
+    """
+    Loop through projects and files one at a time to upload
+    :param api_instance: api instance should have already been initialized
+    :param sequencing_run: sequencing run to upload
+    :param run_id: run id to use for upload
+    :return:
+    """
+    # loop through projects
+    for project in sequencing_run.project_list:
+        # loop through samples
+        for sample in project.sample_list:
+            logging.info("Uploading to Sample {} on Project {}".format(sample.sample_name, project.id))
+            # upload files
+            api_instance.send_sequence_files(sequence_file=sample.sequence_file,
+                                             sample_name=sample.sample_name,
+                                             project_id=project.id,
+                                             upload_id=run_id)
+
+
+def _run_upload_threadpool(api_instance, sequencing_run, run_id, max_threads):
+    """
+    Starts a threading pool with thread count defined in config. Threadpool then uploads samples simultaneously
+    :param api_instance: api instance should have already been initialized
+    :param sequencing_run: sequencing run to upload
+    :param run_id: run id to use for upload
+    :param max_threads: number of threads to use
+    :return:
+    """
+
+    def _run_upload(data_dict, sequence_run_id):
+        """
+        Do the data upload
+        :param data_dict: dictionary containing the "project" and "sample"
+        :param sequence_run_id: run id to upload with
+        :return:
+        """
+        project_data = data_dict["project"]
+        sample_data = data_dict["sample"]
+        logging.debug("Processing queue item for project {} sample {}".format(project_data.id,
+                                                                              sample_data.sample_name))
+
+        api_instance.send_sequence_files(sequence_file=sample_data.sequence_file,
```

Comment:
Getting some odd effects with uploading files in the gui.  I have 4 threads, uploading normal sized files.  It's sometimes getting to about 50% progress, jumping to 100%, then back to the middle and continuing uploading with a green progress bar.  Any idea what that's about?

![image](https://user-images.githubusercontent.com/8951708/76562985-80510a00-6474-11ea-8416-f4353a94d321.png)

Reply:
They green bar at 50% is something I've seen a few times (before adding multithreading) but just didn't get around to making an issue for it. I think it's been there since we added the GUI but only happens rarely. I think that sometimes when completing the first file of a pair it will change the bars colour to green.

Reply:
Yeah I think the issue suffices for now.  Not really a functional error, just something odd.  Thanks,